### PR TITLE
feat: add events to voting v2

### PR DESCRIPTION
**Pull request by md0x**
Signed-off-by: Pablo Maldonado <pablo@umaproject.org>

**Motivation**

Add events and update old ones to the DVM2.0 contracts

**Summary**

- Added events to VotingV2
- Added events to Staker

**Testing**

Check a box to describe how you tested these changes and list the steps for reviewers to test.

- [ ]  Ran end-to-end test, running the code as in production
- [ ]  New unit tests created
- [ ]  Existing tests adequate, no new tests required
- [x]  All existing tests pass
- [ ]  Untested

**Issue(s)**

Fixes #XXXX

### CODE REVIEW

File: packages/core/contracts/oracle/implementation/Staker.sol [hunk 1]
```diff
@@ -112,13 +137,16 @@ contract Staker is StakerInterface, Ownable, Testable {
 
     // Calculate the reward per token based on last time the reward was updated.
     function _updateReward(address voterAddress) internal {
-        rewardPerTokenStored = rewardPerToken();
-        lastUpdateTime = getCurrentTime();
+        uint256 currentTime = getCurrentTime();
+        uint256 newRewardPerToken = rewardPerToken();
```
Comment (md0x) [inline packages/core/contracts/oracle/implementation/Staker.sol:141]: added this aux variables to not read from storage

File: packages/core/contracts/oracle/implementation/VotingV2.sol [hunk 2]
```diff
@@ -351,6 +373,7 @@ contract VotingV2 is
         if (slash + int256(voterStake.cumulativeStaked) > 0)
             voterStake.cumulativeStaked = uint256(int256(voterStake.cumulativeStaked) + slash);
         else voterStake.cumulativeStaked = 0;
+        emit VoterSlashed(voterAddress, slash);
```
Comment (md0x) [inline packages/core/contracts/oracle/implementation/VotingV2.sol:376]: I haven't added event for `_updateAccountSlashingTrackers`, I'm hesitant about it as we already have the `VoterSlashed`. They could be reasons for adding a new event for `_updateAccountSlashingTrackers` for off-chain calculations purposes.

**Review by mrice32 — APPROVED** [review]
Looks great! Only minor comments!

File: packages/core/contracts/oracle/implementation/Staker.sol [hunk 3]
```diff
@@ -85,6 +106,8 @@ contract Staker is StakerInterface, Ownable, Testable {
             voterStake.unstakeTime = 0;
             votingToken.transfer(msg.sender, tokensToSend);
         }
+
+        emit StakeExecuted(msg.sender, tokensToSend);
```
Comment (chrismaree) [inline packages/core/contracts/oracle/implementation/Staker.sol:110]: same nit: rename to `ExecutedUnstake`. Include the stakers `voterStake.cumulativeStaked` as well to see how much was left after the unstake.

File: packages/core/contracts/oracle/implementation/Staker.sol [hunk 4]
```diff
@@ -98,6 +121,8 @@ contract Staker is StakerInterface, Ownable, Testable {
             voterStake.outstandingRewards = 0;
             require(votingToken.mint(msg.sender, tokensToMint), "Voting token issuance failed");
         }
+        emit RewardsWithdrawn(msg.sender, tokensToMint);
```
Comment (chrismaree) [inline packages/core/contracts/oracle/implementation/Staker.sol:124]: `WithdrawnRewards`

File: packages/core/contracts/oracle/implementation/Staker.sol [hunk 5]
```diff
@@ -138,9 +166,11 @@ contract Staker is StakerInterface, Ownable, Testable {
     function setEmissionRate(uint256 _emissionRate) public onlyOwner {
         _updateReward(address(0));
         emissionRate = _emissionRate;
+        emit NewEmissionRate(emissionRate);
```
Comment (chrismaree) [inline packages/core/contracts/oracle/implementation/Staker.sol:169]: if you want to name it like this then go `SetNewEmissionRate`

File: packages/core/contracts/oracle/implementation/VotingV2.sol [hunk 6]
```diff
@@ -233,6 +239,22 @@ contract VotingV2 is
...
+
+    event GatPercentageChanged(FixedPoint.Unsigned newGatPercentage);
+
+    event SlashingLibraryChanged(address newAddress);
+
+    event SpamDeletionProposalBondChanged(uint256 newBond);
+
+    event VoterSlashed(address indexed voter, int256 numTokens);
+
+    event CumulativeSlashingTrackersUpdated(uint256 lastRequestIndexConsidered, uint256 priceRequestIdsLength);
```
Comment (chrismaree) [inline packages/core/contracts/oracle/implementation/VotingV2.sol:252]: we'll need to re-consider the props here but looks good for now.

File: packages/core/contracts/oracle/implementation/VotingV2.sol (hunk 2 above)
Comment (chrismaree) [reply packages/core/contracts/oracle/implementation/VotingV2.sol:376]: yup. this is the only place where voters' balance actually go down. we should probs include `voterStake.cumulativeStaked` as well.

File: packages/core/contracts/oracle/implementation/VotingV2.sol [hunk 7]
```diff
@@ -395,6 +418,8 @@ contract VotingV2 is
 
             lastRequestIndexConsidered = i + 1;
         }
+
+        emit CumulativeSlashingTrackersUpdated(lastRequestIndexConsidered, priceRequestIdsLength);
```
[the remark this answers is not in the document]
Comment (chrismaree) [reply packages/core/contracts/oracle/implementation/VotingV2.sol:422]: ye, we might want to remove this one for now and come back to it a bit later

File: packages/core/contracts/oracle/implementation/VotingV2.sol [hunk 8]
```diff
@@ -427,6 +452,8 @@ contract VotingV2 is
         bytes32 identifier = SpamGuardIdentifierLib._constructIdentifier(proposalId);
 
         _requestPrice(identifier, currentTime, "", true);
+
+        emit SpamRequestsSignaled(proposalId, msg.sender, spamRequestIndices);
```
Comment (chrismaree) [inline packages/core/contracts/oracle/implementation/VotingV2.sol:456]: nit: `SignaledRequestsAsSpamForDeletion`

File: packages/core/contracts/oracle/implementation/VotingV2.sol [hunk 9]
```diff
@@ -464,15 +491,18 @@ contract VotingV2 is
 
             // Return the spamDeletionProposalBond.
             votingToken.transfer(spamDeletionProposals[proposalId].proposer, spamDeletionProposalBond);
+            emit SpamRequestResolved(proposalId, true);
```
Comment (chrismaree) [inline packages/core/contracts/oracle/implementation/VotingV2.sol:494]: `ExecutedSpamDeletion`

File: packages/core/contracts/oracle/implementation/Staker.sol [hunk 10]
```diff
@@ -43,7 +43,7 @@ contract Staker is StakerInterface, Ownable, Testable {
      ****************************************/
 
     event Staked(
-        address indexed staker,
+        address indexed voter,
```
Comment (md0x) [inline packages/core/contracts/oracle/implementation/Staker.sol:46]: I renamed to `voter` to be consistent with the naming

File: packages/core/contracts/oracle/implementation/Staker.sol [hunk 11]
```diff
@@ -77,6 +129,13 @@ contract Staker is StakerInterface, Ownable, Testable {
         voterStakes[msg.sender].pendingUnstake = amount;
         voterStakes[msg.sender].activeStake -= amount;
         voterStakes[msg.sender].unstakeTime = getCurrentTime() + unstakeCoolDown;
+        emit RequestedUnstake(
```
Comment (chrismaree) [inline packages/core/contracts/oracle/implementation/Staker.sol:132]: it might be better for this method to define a `VoterStake storage voterStake = voterStakes[msg.sender];` and then reference this. you can leave it for now but just looking at some of the changes you have here

**Review by chrismaree — APPROVED** [review]
looks good to me! some unit tests will be useful. it might be good to create _one_ test for all events that tests the full lifecycle over all the actions and events.

File: packages/core/contracts/oracle/implementation/Staker.sol (hunk 11 above)
Comment (md0x) [reply packages/core/contracts/oracle/implementation/Staker.sol:132]: Well remarked, I finally decided to focus on events and leave the storage reading optimisations for a second pass. But there is definitely room for improvement here.

File: packages/core/test/oracle/VotingV2.js [hunk 12]
```diff
@@ -2663,6 +2663,93 @@ describe("VotingV2", function () {
     await voting.methods.setDelegator(ZERO_ADDRESS).send({ from: rand });
     assert.equal(await voting.methods.delegateToStaker(rand).call(), ZERO_ADDRESS);
   });
+
+  it("Emit events", async function () {
```
Comment (chrismaree) [inline packages/core/test/oracle/VotingV2.js:2667]: this feels somewhat similar to the events here: https://github.com/UMAprotocol/protocol/blob/94d53ffc7fdafee7b5a4443d83dd50d02c0e72e2/packages/core/test/oracle/VotingV2.js#L887. do you want to shorten this to _just_ be `Emit StakeEvents` or perhaps merge this with the tests there? Also, for the staking specific tests it might be better to pull some of this into this location: https://github.com/UMAprotocol/protocol/blob/master/packages/core/test/oracle/Staker.ts

sorry that I'm making you move stuff around...Just would ideally not like too much duplication in the tests.

Comment (md0x) [reply packages/core/test/oracle/VotingV2.js]: No problem! I missed that file and existing test! I have reorganised the code reuse the existing tests and moved the staking ones to the corresponding file.
